source/command-logging-and-monitoring/command-logging-and-monitoring.rst

```diff
@@ -0,0 +1,615 @@
+     * The calculated value MUST be the time to send the message and receive the reply from the server
+     * and MAY include BSON serialization and/or deserialization. The name can imply the units in which the
+     * value is returned, i.e. durationMS, durationNanos.
+     */
+    duration: Int64;
+
+    /**
+     * Returns the command name.
+     */
+    commandName: String;
+
+    /**
+     * Returns the failure. Based on the language, this SHOULD be a message string, exception
+     * object, or error document.
+     */
+    failure: String,Exception,Document;
+
+    /**
+     * Returns the client generated request id.
+     */
+    requestId: Int64;
+
+    /**
+     * Returns the driver generated operation id. This is used to link events together such
+     * as bulk write operations. OPTIONAL.
+     */
+    operationId: Int64;
+
+    /**
+     * Returns the connection id for the command. For languages that do not have this,
+     * this MUST return the driver equivalent which MUST include the server address and port.
+     * The name of this field is flexible to match the object that is returned from the driver.
+     */
+    connectionId: ConnectionId;
+
+    /**
+     * Returns the server connection id for the command. The server connection id is distinct from
+     * the connection id and is returned by the hello or legacy hello response as "connectionId"
+     * from the server on 4.2+. Drivers MAY use a wider type to represent the server connection ID
+     * value, but the server's behavior is to return an Int32.
+     */
+    serverConnectionId: Optional<Int32>;
+
+    /**
+     * Returns the service id for the command when the driver is in load balancer mode.
+     * For drivers that wish to include this in their ConnectionId object, this field is
+     * optional.
+     */
+    serviceId: Optional<ObjectId>;
+  }
+
+------------
+Log Messages
+------------
+Please refer to the `logging specification <../logging/logging.rst>`_ for details on logging implementations in general, including log levels, log
+components, and structured versus unstructured logging.
+
+Drivers MUST support logging of command information via the following types of log messages. These messages MUST be logged at ``Debug`` level and use
+the ``command`` log component.
+
+The log messages are intended to match the information contained in the events above. Drivers MAY implement command logging
+support via an event subscriber if it is convenient to do so.
+
+The types used in the structured message definitions below are demonstrative, and drivers MAY use similar types instead so long as the information
+is present (e.g. a double instead of an integer, or a string instead of an integer if the structured logging framework does not support numeric types.) 
+
+For the string representations of messages defined below, for any optional field that is not available, drivers MAY either substitute the value with a
```

Comment:
I can see arguments for either of these. on the one hand, it might wrongly concern users to see a value included in log messages is always null (for example if they are on < 4.2 and serverConnectionId is always null, or they aren't doing any bulk writes and so the operationId is always missing). also, it will make messages longer/arger if we always include the extra value.

on the other hand, in some cases it may give a more explicit indicator of a problem when debugging to say that something is null. a TSE or user reading a log won't necessarily know that a missing value means the driver considers the value to be null or that the value should have been included at all.

any thoughts?

Reply:
I prefer to leave them out.  On balance, it's more confusing to include a null for a value that the driver knows is always going to be null in some situations, than to leave it out and risk confusing everyone reading.  I've seen something similar with the C# drivers's representation of host names in server selection timeout exception messages.  I think the ipV6 address is always printed as null even though it rarely exists.  It always confuses me.